### ISSUE TITLE
Implement a way to generate classes or types from JSON Schema definitions

### ISSUE THREAD

**Issue by derberg** [asyncapi/generator#216]
**Description**

Integrate some solutions like https://quicktype.io/ so you can easily get from the generator models for any programming language you need. Make research and proposal first, if needed.

Support as many languages as possible, depending on the final solution we will choose. First on the lists are:
- The ones that are already used in templates
- Rust as there is a need for it #53 

**Hint** Some prototyping was already done as a part of an issue where Java POJO generation was requested #159 

**Reason**

Make the generator more feature-rich for code generation use cases.

**Comment by jonaslagoni** [comment]
Think this is blocked by https://github.com/asyncapi/generator/issues/239

**Comment by Tenischev** [comment]
Here is POJO model in java-spring template 
https://github.com/asyncapi/java-spring-template/blob/master/template/src/main/java/com/asyncapi/model/%24%24schema%24%24.java
Currently this model has following features:
- usage of validation annotations
- usage of jackson annotations
- generation of enums
- generation of javadocs based on description for field

**Comment by derberg** [comment]
I research a lot and haven't found any good competition against quicktype that we could use in node

**Comment by derberg** [comment]
I research a lot and haven't found any good competition against quicktype that we could use in node. But the learning curve is steep as hell as there are basically no docs and code base is TS. This comment summarizes my concerns https://github.com/quicktype/quicktype/issues/1293#issuecomment-523263652

But I do not want to give up and will try to dig into more details and create a sample project to check how much we can influence the quicktype output.

**Comment by derberg** [comment]
quicktype function options:
- https://github.com/quicktype/quicktype/blob/master/src/quicktype-core/Run.ts#L119-L169
- https://github.com/quicktype/quicktype/blob/master/src/quicktype-core/Run.ts#L36-L42

quicktype returns the following type: https://github.com/quicktype/quicktype/blob/master/src/quicktype-core/Source.ts#L158

😅

**Comment by derberg** [comment]
I made basic stuff here https://github.com/derberg/asyncapi-quicktype

By default, from such a schema from streetlights example:
```
{
  type: 'object',
  properties: {
    percentage: {
      type: 'integer',
      description: 'Percentage to which the light should be dimmed to.',
      minimum: 0,
      maximum: 100
    },
    sentAt: {
      type: 'string',
      format: 'date-time',
      description: 'Date and time when the message was sent.',
      'x-parser-schema-id': 'sentAt'
    }
  },
  'x-parser-schema-id': 'dimLightPayload'
}
```

I got the following Java:
```
/ Converter.java

// To use this code, add the following Maven dependency to your project:
//
//
//     com.fasterxml.jackson.core     : jackson-databind          : 2.9.0
//     com.fasterxml.jackson.datatype : jackson-datatype-jsr310   : 2.9.0
//
// Import this package:
//
//     import io.quicktype.Converter;
//
// Then you can deserialize a JSON string with
//
//     DimLightPayload data = Converter.fromJsonString(jsonString);

package io.quicktype;

import java.io.IOException;
import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import java.util.*;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.OffsetTime;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.time.temporal.ChronoField;

public class Converter {
    // Date-time helpers

    private static final DateTimeFormatter DATE_TIME_FORMATTER = new DateTimeFormatterBuilder()
            .appendOptional(DateTimeFormatter.ISO_DATE_TIME)
            .appendOptional(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
            .appendOptional(DateTimeFormatter.ISO_INSTANT)
            .appendOptional(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SX"))
            .appendOptional(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ssX"))
            .appendOptional(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
            .toFormatter()
            .withZone(ZoneOffset.UTC);

    public static OffsetDateTime parseDateTimeString(String str) {
        return ZonedDateTime.from(Converter.DATE_TIME_FORMATTER.parse(str)).toOffsetDateTime();
    }

    private static final DateTimeFormatter TIME_FORMATTER = new DateTimeFormatterBuilder()
            .appendOptional(DateTimeFormatter.ISO_TIME)
            .appendOptional(DateTimeFormatter.ISO_OFFSET_TIME)
            .parseDefaulting(ChronoField.YEAR, 2020)
            .parseDefaulting(ChronoField.MONTH_OF_YEAR, 1)
            .parseDefaulting(ChronoField.DAY_OF_MONTH, 1)
            .toFormatter()
            .withZone(ZoneOffset.UTC);

    public static OffsetTime parseTimeString(String str) {
        return ZonedDateTime.from(Converter.TIME_FORMATTER.parse(str)).toOffsetDateTime().toOffsetTime();
    }
    // Serialize/deserialize helpers

    public static DimLightPayload fromJsonString(String json) throws IOException {
        return getObjectReader().readValue(json);
    }

    public static String toJsonString(DimLightPayload obj) throws JsonProcessingException {
        return getObjectWriter().writeValueAsString(obj);
    }

    private static ObjectReader reader;
    private static ObjectWriter writer;

    private static void instantiateMapper() {
        ObjectMapper mapper = new ObjectMapper();
        mapper.findAndRegisterModules();
        mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        SimpleModule module = new SimpleModule();
        module.addDeserializer(OffsetDateTime.class, new JsonDeserializer<OffsetDateTime>() {
            @Override
            public OffsetDateTime deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException {
                String value = jsonParser.getText();
                return Converter.parseDateTimeString(value);
            }
        });
        mapper.registerModule(module);
        reader = mapper.readerFor(DimLightPayload.class);
        writer = mapper.writerFor(DimLightPayload.class);
    }

    private static ObjectReader getObjectReader() {
        if (reader == null) instantiateMapper();
        return reader;
    }

    private static ObjectWriter getObjectWriter() {
        if (writer == null) instantiateMapper();
        return writer;
    }
}

// DimLightPayload.java

package io.quicktype;

import com.fasterxml.jackson.annotation.*;
import java.time.OffsetDateTime;

public class DimLightPayload {
    private Long percentage;
    private OffsetDateTime sentAt;

    /**
     * Percentage to which the light should be dimmed to.
     */
    @JsonProperty("percentage")
    public Long getPercentage() { return percentage; }
    @JsonProperty("percentage")
    public void setPercentage(Long value) { this.percentage = value; }

    /**
     * Date and time when the message was sent.
     */
    @JsonProperty("sentAt")
    public OffsetDateTime getSentAt() { return sentAt; }
    @JsonProperty("sentAt")
    public void setSentAt(OffsetDateTime value) { this.sentAt = value; }
}
```


Now I will explore how this output can be modified

**Comment by derberg** [comment]
I can confirm using quicktype you have a control over the generated code. 
Example here: https://github.com/derberg/asyncapi-quicktype/blob/master/src/customJava.js#L17


Output with my custom annotation:

```java
package io.quicktype;

import java.time.OffsetDateTime;

@ThisIsMySuperCustomAnnotation
public class DimLightPayload {
    private Long percentage;
    private OffsetDateTime sentAt;

    /**
     * Percentage to which the light should be dimmed to.
     */
    public Long getPercentage() { return percentage; }
    public void setPercentage(Long value) { this.percentage = value; }

    /**
     * Date and time when the message was sent.
     */
    public OffsetDateTime getSentAt() { return sentAt; }
    public void setSentAt(OffsetDateTime value) { this.sentAt = value; }
}
```

**Comment by derberg** [comment]
My view is that we should not add it to the generator right before major release. Why:
- feature is too big and complex. In Poland we say that such features "musza sie wygrzac", which means they need time to worm up. It is super hard to say if whatever API we will agree up on now will really stay like this
- we can go small and just allow simple generation and they add support for customizations, but this can mean breaking changes that are hard to foresee. Maybe I made it running but I've spent on it less than 2 days doing other things and hard to say what are the other options and if what I do is good
- because of no docs in the project, I think that the only option might be introducing TypeScript in the generator to leverage on IDEs support and nice suggestions and autocompletion

As if it comes to "should we use quicktype or not" - yes we should, the learning curve is steep but the stuff that we need we can well explain in our docs to make it easier for template devs.

Now, these are the options to integrate:
1. The easiest one is through the hooks. We could provide a hook in the generator-hooks library that would be responsible for generation. We would only have to introduce in the template configuration a new prop, like "language" and we would use this info in the hook to decide in what language to generate models. Other things like package names could be parametrized with default values support.
Hooks would work for basic support, but not for customizations as we would have to somehow be able to pass into the hook a new extended TargetLanguage.
So the advantage is we can do it without touching the generator API, at any change without breaking changes. But to start supporting customizations we would probably have to break below:
    ```
    "hooks": {
        "@asyncapi/generator-hooks": "hookFunctionName"
      }
    ```
Also quicktype itself is super powerful with complex options available and with hooks, I have a feeling we might limit possibilities.
2. More work would be to add it as a completely new feature, like core, and not something that is ran at hooks. This would be a new feature, better to advertise, better to notice by template devs (rather than a hidden hook).
    ```
    "quicktype-models": {
        "lang": "kotlin",
        "options": Object,
        "langOptions": Object,
        "customTargetLanguage: Module
      }
    ```
I would not hide the name but clearly write as in my example that it is based on quicktype.
3. File templates? Have new `$$schema-quicktype$$ where you would have some extension points where you would specify lang, options and custom classes for quicktype

I think 2 is nicest. 
I'm super interested in your opinions.

**Comment by fmvilas** [comment]
Great analysis mate. I'd make it a core feature and agree to make it explicit that we're using quicktype. And yes, it's too risky to add it to this major release. We can roll it out afterward and happy to bump major version if needed.

**Comment by derberg** [comment]
work on research how to do this is [here](https://github.com/asyncapi/shape-up-process/issues/21) (better view is [here](https://shapeup.asyncapi.io/cycles/6054195?bet=21))

**Comment by jonaslagoni** [comment]
Solution details here: https://github.com/asyncapi/shape-up-process/issues/43
Repository for the library here: https://github.com/asyncapi/generator-model-sdk/
